File: LeetCode/HeapPriorityQueue/python/kthLargestElement.py
from dataclasses import dataclass
from enum import Enum
import heapq
from typing import List

class Result(Enum):
    PASS = "\033[92mPASS\033[00m"
    FAIL = "\033[91mFAIL\033[00m"

@dataclass
class findKthLargestRecord:
    nums: List[int]
    k: int
    expected: float

class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        h_min: heapq = heapq
        heap: List[int] = []

        for el in nums[0:k]:
            h_min.heappush(heap, el)

        for el in nums[k:]:
            if el > heap[0]:
                h_min.heappop(heap)
                h_min.heappush(heap, el)

        return heap[0]

    @staticmethod
    def testSolution(record: findKthLargestRecord) -> None:
        print(f"input:\tnums: {record.nums}")
        print(f"\tk: {record.k}")
        print(f"expected: {record.expected}")

        res: int = Solution().findKthLargest(record.nums, record.k)

        print(f"result: {res}")
        print(Result.PASS.value if res == record.expected else Result.FAIL.value)

if __name__ == "__main__":
    records: findKthLargestRecord = [
        findKthLargestRecord([3,2,1,5,6,4], 2, 5 ),
        findKthLargestRecord([3,2,3,1,2,4,5,5,6], 4, 4)
    ]

    for i, record in enumerate(records):
        print(f"# Test case {i + 1}")
        Solution.testSolution(record)
        print("-" * 50)

File: LeetCode/HeapPriorityQueue/java/kthLargestElement/kthLargestElement.java
package LeetCode.HeapPriorityQueue.java.kthLargestElement;

import java.util.Arrays;
import java.util.List;
import java.util.PriorityQueue;
import java.util.stream.Collectors;

enum testResult {
    PASS("\u001B[32mPASS\u001B[0m"),
    FAIL("\u001B[31mFAIL\u001B[0m");

    private final String value;

    testResult(String value) {
        this.value = value;
    }

    public String getValue() {
        return this.value;
    }
}

record kthLargestElementRecord(int[] nums, int k, int expected) {
}

class kthLargestElement {
    public int findKthLargest(int[] nums, int k) {
        PriorityQueue<Integer> hMin = new PriorityQueue<>();

        // List<Integer> list = new ArrayList<>();

        // for( int i = 0 ; i < nums.length ; i++ ){
        // list.add(nums[i]);
        // }

        List<Integer> list = Arrays.stream(nums)
                .boxed()
                .collect(Collectors.toList());

        list.subList(0, k).stream()
                .forEach(n -> hMin.add(n));

        list.subList(k, nums.length).stream()
                .forEach(n -> {
                    if (n > hMin.peek()) {
                        hMin.poll();
                        hMin.add(n);
                    }
                });

        // System.out.println(hMin);

        return hMin.peek();
    }

    public static void main(String[] args) {

        kthLargestElementRecord[] records = new kthLargestElementRecord[] {
                new kthLargestElementRecord(new int[] { 3, 2, 3, 1, 2, 4, 5, 5, 6 }, 4, 4),
                new kthLargestElementRecord(new int[] { 3, 2, 1, 5, 6, 4 }, 2, 5)
        };

        int i = 1;
        for (kthLargestElementRecord record : records) {
            System.out.println("# Test case " + i++);
            kthLargestElement.testSolution(record);
            System.out.println("-".repeat(50));
        }
    }

    private static void testSolution(kthLargestElementRecord record) {
        System.out.println("input:\tnums: " + Arrays.toString(record.nums()));
        System.out.println("\tk: " + record.k());
        System.out.println("expected: " + record.expected());

        int res = new kthLargestElement().findKthLargest(record.nums(), record.k());
        System.out.println("result: " + res);
        System.out.println(res == record.expected() ? testResult.PASS.getValue() : testResult.FAIL.getValue());

    }
}